source/dub/generators/build.d

```diff
@@ -624,98 +635,232 @@ struct Watcher
+				return getAncestor(paths, accum, 1);
+			} else {
+				if(paths.all!(p => p.startsWith(accum)) && accum.length() < paths[0].length())// && accum != paths[0])
+					return getAncestor(paths, accum ~ paths[0][cdepth], cdepth + 1);
+				else
+					return accum.parentPath();
+			}
+		}
+		// Gets the most specific Ancestor Path a list of absolute paths:
+		private string getAncestor(string[] paths) {
+			auto ps = paths.map!(path => Path(path)).array; // lazy better?
+			return to!string(getAncestor(ps)).replace("/", "\\");
+		}
+		// returns true if watched file has changed:
+		bool processBuffer(void* buf) {
+			struct FILE_NOTIFY_INFORMATION {
+				uint NextEntryOffset;
+				uint Action;
+				uint FileNameLength;
+				wchar[0] FileName; // variable length struct
+			}
+			logInfo("Processing buffer.");
+			int nextOffset = 0;
+			FILE_NOTIFY_INFORMATION* info;
+			do {
+				info = cast(FILE_NOTIFY_INFORMATION*)(buf + nextOffset);
+				auto ws = (cast(wchar*)info.FileName)[0..info.FileNameLength/wchar.sizeof];
+				string fname = to!string(to!wstring(ws)); // TODO: make this better?
+				string fpath = joinPath(dir, fname).replace("/", "\\"); // why does vibe.d do this?
+				if(paths.canFind(fpath) && info.Action == FILE_ACTION_MODIFIED) {
+				   return true; 
+				}
+				// TODO: Check if this is actually a correct loop.
+				nextOffset += info.NextEntryOffset; // NextEntryOffset = 0 when it's the last. 
+				logInfo("Unwatched file event: " ~ fpath);
+			} while (info.NextEntryOffset != 0);
+			return false;
+		}
+		void createFileHandle() {
+			if(hDir != INVALID_HANDLE_VALUE)
+				return;
+			dir = getAncestor(paths);
+			logInfo("Determined to watch: " ~ dir);
+			hDir = CreateFileW(toUTF16z(dir),
+							   FILE_LIST_DIRECTORY, // required for reading changes
+							   FILE_SHARE_READ | FILE_SHARE_WRITE, // No delete b/c
+							   // otherwise user can delete the directory.
+							   null, // security attr.
+							   OPEN_EXISTING,
+							   FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
+							   null);
+			if(hDir == INVALID_HANDLE_VALUE)
+				assert(0, "Couldn't open directory.");
+		}
+		void wait(Pid pid)
 		{
-			fd_set rfds=void;
-			FD_ZERO(&rfds);
-			FD_SET(eventfd, &rfds);
+			createFileHandle();
+			logInfo("Watching for changes in the files or process exit.");
+			immutable buff_siz = 4096;
+			void* buffer = cast(void*)(new byte[buff_siz]);
+			uint bytesReturned = 0; // completely useless
+			OVERLAPPED ovlap;  // also useless
+			uint result;
+			do
+			{
+				if(!tryWait(pid).terminated) {
+					auto initial = ReadDirectoryChangesW(hDir, buffer, buff_siz,
+														 true, // watch subdirs
+														 FILE_NOTIFY_CHANGE_LAST_WRITE,
+														 &bytesReturned,
+														 &ovlap, cast(void*)&_CompletionRoutine);
+					assert(initial, "winapi ReadDirectoryChangesW");
+					HANDLE hProc = pid.osHandle();
+					result = WaitForMultipleObjectsEx(1, &hProc, true, INFINITE, true);
```

Comment:
This API call is key because it allows the watcher thread to wait for two things at once: the process exiting and any file changes.